modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java

```diff
@@ -913,6 +914,29 @@ public static void inBusyLock(IgniteSpinBusyLock busyLock, Runnable fn) {
         }
     }
 
+    /**
+     * Method that runs the provided {@code fn} in {@code busyLock}.
+     *
+     * @param <T> Type of returned value from {@code fn}.
+     * @param busyLock Component's busy lock.
+     * @param fn Function to run.
+     * @return Future returned from the {@code fn}, or future with the {@link NodeStoppingException} if
+     *      {@link IgniteSpinBusyLock#enterBusy()} failed or with runtime exception/error while executing the {@code fn}.
+     */
+    public static <T> CompletableFuture<T> inBusyLockAsync(IgniteSpinBusyLock busyLock, Supplier<CompletableFuture<T>> fn) {
```

Comment:
I seriously doubt this method because it looks very unclear in the code and provokes closure creation. A code with a plain checking lock looks easier to read.
Moreover, I also do not advise abusing the methods (inBusyLock) above for the same reason.
But it is your choose.

Reply:
Thanks for the advice, I'll leave it.